uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -289,318 +115,429 @@
 //! [`Waker`]: https://doc.rust-lang.org/std/task/struct.Waker.html
 //! [`RawWaker`]: https://doc.rust-lang.org/std/task/struct.RawWaker.html
 
-use crate::{ffi::rustcalls::rust_call_with_out_status, FfiConverter, RustCallStatus};
+use crate::{
+    rust_call_with_out_status, schedule_raw, FfiConverter, FfiDefault, ForeignExecutor,
+    ForeignExecutorHandle, RustCallStatus,
+};
 use std::{
-    ffi::c_void,
+    cell::UnsafeCell,
     future::Future,
-    mem::{ManuallyDrop, MaybeUninit},
     panic,
     pin::Pin,
-    sync::Arc,
+    sync::{
+        atomic::{AtomicU32, Ordering},
+        Arc,
+    },
     task::{Context, Poll, RawWaker, RawWakerVTable, Waker},
 };
 
-/// `RustFuture` represents a [`Future`] that can be sent over FFI safely-ish.
+/// Callback that we invoke when a `RustFuture` is ready.
 ///
-/// See the module documentation to learn more.
-#[repr(transparent)]
-pub struct RustFuture<T>(Pin<Box<dyn Future<Output = T> + 'static>>);
-
-impl<T> RustFuture<T> {
-    /// Create a new `RustFuture` from a regular `Future`.
-    pub fn new<F>(future: F) -> Self
-    where
-        F: Future<Output = T> + 'static,
-    {
-        Self(Box::pin(future))
-    }
+/// The foreign code passes a pointer to one of these callbacks along with an opaque data pointer.
+/// When the future is ready, we invoke the callback.
+pub type FutureCallback<T> =
+    extern "C" fn(callback_data: *const (), result: T, status: RustCallStatus);
 
-    /// Create a new `RustFuture` from a Tokio `Future`. It needs an
-    /// intermediate compatibility layer to be handlded as a regular Rust
-    /// `Future`.
-    #[cfg(feature = "tokio")]
-    pub fn new_tokio<F>(future: F) -> Self
-    where
-        F: Future<Output = T> + 'static,
-    {
-        Self(Box::pin(async_compat::Compat::new(future)))
-    }
+/// Future that the foreign code is awaiting
+///
+/// RustFuture is always stored inside a Pin<Arc<>>.  The `Arc<>` allows it to be shared between
+/// wakers and Pin<> signals that it must not move, since this would break any self-references in
+/// the future.
+pub struct RustFuture<F, T, UT>
+where
+    // The future needs to be `Send`, since it will move to whatever thread the foreign executor
+    // chooses.  However, it doesn't need to be `Sync', since we don't share references between
+    // threads (see do_wake()).
+    F: Future<Output = T> + Send,
```

Comment:
I've just bumped to the version of main with this PR merged and now I'm seeing a lot of "future returned by `FOO` is not `Send`" on my `[uniffi::export]`s. e.g. from things such as uses of `itertools::IntoChunks` within the function (and an async call using a chunk) and similar things.

I've not worked through all the cases we are seeing but some of them are looking like they will tricky to resolve, especially in a code base which also targets wasm (which is inherently single threaded so various types, such as `JsValue` and event `JsFuture` aren't `Send` etc). and it looks like our use of the Actix test stuff is going to run into trouble too since e.g. `actix_web::dev::Service` isn't `Send`.

I'm unsure if this is actually a new requirement or if it's always been this way under the hood but wasn't made visible to Rust until now, so the compiler wasn't catching it. Can we see anyway that the requirement can be relaxed? I assume it's tricky (or even impossible) due to the nature of the foreign executor?

Reply:
This is a tricky one.  When I wrote the code I was under the impression that it was always a requirement and we should have had the bound all along.  I think that's true in the general case, but I also think you're correct that with certain setups it won't be needed.

I made https://github.com/mozilla/uniffi-rs/issues/1538 to discuss this more.